src/dependency-manager/src/configs/graph/index.ts

```diff
@@ -0,0 +1,261 @@
+/* eslint-disable no-empty */
+import { EnvironmentVault } from '../../environment-config/base';
+import VaultManager from '../../vault-manager';
+import { BaseDnsConfig, BaseEnvironmentConfig } from '../base-configs/environment-config';
+import { BaseParameterValueConfig, BaseParameterValueFromConfig, BaseServiceConfig, BaseValueFromDependencyConfig, BaseValueFromVaultConfig } from '../base-configs/service-config';
+import { EnvironmentBuilder } from '../environment.builder';
+import { ServiceBuilder } from '../service.builder';
+
+interface GraphEnrichmentOptions {
+  getConfig: (ref: string) => Promise<BaseServiceConfig>;
+  getRegistryImage: (ref: string) => string;
+  getHostAssignment: (config: BaseServiceConfig, interface_key: string) => Promise<string>;
+  getPortAssignment: (config: BaseServiceConfig, interface_key: string) => Promise<number>;
+  debug?: boolean;
+  skip_nodes?: string[];
+}
+
+class GraphNotBuiltError extends Error {
+  message = 'Please build and validate the graph before applying changes';
+}
+
+class ValueFromDependencyError extends Error {
+  constructor(
+    service_name: string,
+    dependency_name: string,
+    parameter_key: string
+  ) {
+    super();
+    this.message = `${service_name} is trying to populate the parameter, ${parameter_key}, from the service ${dependency_name}, but it was not installed as a dependency.`;
+  }
+}
+
+export class EnvironmentGraph {
+  private _is_built = false;
+  private _base_config: BaseEnvironmentConfig;
+  private _enriched_config: BaseEnvironmentConfig;
+
+  constructor(config: BaseEnvironmentConfig) {
```

Comment:
I haven't found a case for stepwise yet actually. I just changed this to use a static `build()` method instead that's async and loads everything up. It still takes an environment config (which should be built and optionally validated in a separate step), but once you are seeking a graph it makes sense that it will always be built/enriched.